### DISCUSSION TITLE
Guidance on how to implement padding.width and window.(in)active.title.separator.color

### DISCUSSION THREAD

**Discussion by ludgie** [labwc/labwc#1300] [Q&A]
Hello labwc developers,

I've tried for a few hours on how to implement the padding.width and window.active.title.separator.color and window.inactive.title.separator.color, although I don't know how to continue and I'd like some guidance on how to finish the implementation of these properties for themes.

padding.width: I've set the foundation for the theme engine to detect the property under themerc.
By what I've understood reading the code is that the titlebar buttons are created on the ssd_titlebar_create function under ssd_titlebar.c, so I'm guessing that when the buttons and the title are created with the functions inside ssd_titlebar_create, the padding should be added as well per button and the title. Although I'm probably very wrong.

window.(in)active.title.separator.color: set the foundation for the theme engine to detect the properties under themerc.
Tried to add the separator border with add_scene_rect on ssd_border_create under ssd_border.c, didn't work, although I've understood that there is where borders are created.

A non-functional branch with the current changes can be seen under https://github.com/ludg1e/labwc/tree/ob-ssd-misc

Have a nice day

**Comment by johanmalm** [comment]
Let's start with the horizontal padding.

I think we need to add it here:

https://github.com/labwc/labwc/blob/02aa000b0db211dc3ff5aae6064342008d749b23/src/ssd/ssd_titlebar.c#L96

Probably best to create a new local "button_width" variable and make it SSD_BUTTON_WIDTH + padding.width

Tip: try setting a really big padding width so that it's obvious how it affects the buttons

**Reply by ludgie** [reply]
So are you suggesting add a new argument to add_scene_button, or add ("somehow") the button_width local variable to the int x argument of add_scene_button?

Apologies if I didn't understand well what you said

**Reply by ludgie** [reply]
nvm understood it already

**Reply by ludgie** [reply]
Tried what you've suggested and doesn't seem to work: https://github.com/ludg1e/labwc/commit/8d0d700d6c057c6b9640fb3dbea853ca037975c0
I'm setting the padding.width through themerc
![image](https://github.com/labwc/labwc/assets/77110723/b7a7339f-99b1-42af-a3fb-c19f33080ecf)

![image](https://github.com/labwc/labwc/assets/77110723/0c826eea-a323-4730-a6f0-848da9d3958d)

**Reply by ludgie** [reply]
Setting padding.width at 2000 looks like this:
![image](https://github.com/labwc/labwc/assets/77110723/67a5e45a-c706-462d-b467-073ec7a2b5c1)

**Comment by johanmalm** [comment]
Sorry. Was just looking at the phone earlier and pointed at the wrong bit.

This works:

```
diff --git a/src/ssd/ssd_titlebar.c b/src/ssd/ssd_titlebar.c
index fc28e64..7661471 100644
--- a/src/ssd/ssd_titlebar.c
+++ b/src/ssd/ssd_titlebar.c
@@ -167,23 +167,24 @@ ssd_titlebar_update(struct ssd *ssd)
 	struct ssd_sub_tree *subtree;
 	FOR_EACH_STATE(ssd, subtree) {
 		wl_list_for_each(part, &subtree->parts, link) {
+			int padding = 30;
 			switch (part->type) {
 			case LAB_SSD_PART_TITLEBAR:
 				wlr_scene_rect_set_size(
 					wlr_scene_rect_from_node(part->node),
-					width - SSD_BUTTON_WIDTH * SSD_BUTTON_COUNT,
+					width - SSD_BUTTON_WIDTH * 2,
 					theme->title_height);
 				continue;
 			case LAB_SSD_BUTTON_ICONIFY:
 				if (is_direct_child(part->node, subtree)) {
 					wlr_scene_node_set_position(part->node,
-						width - SSD_BUTTON_WIDTH * 3, 0);
+						width - SSD_BUTTON_WIDTH * 3 - padding * 2, 0);
 				}
 				continue;
 			case LAB_SSD_BUTTON_MAXIMIZE:
 				if (is_direct_child(part->node, subtree)) {
 					wlr_scene_node_set_position(part->node,
-						width - SSD_BUTTON_WIDTH * 2, 0);
+						width - SSD_BUTTON_WIDTH * 2 - padding * 1, 0);
 					struct ssd_button *button =
 						node_ssd_button_from_node(part->node);
 					if (button->toggled) {
```

**Comment by johanmalm** [comment]
Having said that, I think this needs thinking through more...

I suggest some refactoring before adding horizontal padding:

1. De-couple corner pieces from menu + close buttons.
2. Do some design/thinking around what the geometric variables mean. `SSD_BUTTON_WIDTH` is not the same is the width of the icon - but rather just a square box (bigger than the icon) within which the icon is centered. In Openbox the buttons can actually be closer together if no padding is specified.
3. Refactor accordingly to knowedge gained from step (2)
4. Introduce `padding.width`

**Reply by ludgie** [reply]
I guess you're reffering to PRs #1289 and #1047 , plus others...

I'll leave this on halt and continue working on other things
Thanks for the guidance and have a nice day (or night)

**Reply by johanmalm** [reply]
Sorry. Yes, it got rather hard quite quickly. Will get there one day :)
